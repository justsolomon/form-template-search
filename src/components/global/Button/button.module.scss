@import 'styles/variables';
@import 'styles/mixins';

.button {
  @include flex-center();

  height: 40px;
  padding: 0 16px;
  border: none;
  cursor: pointer;
  color: #fff;
  font-size: $font-size--md;
  line-height: $font-size--lg;
  font-weight: $font-weight--bold;
  border-radius: $border-radius--sm;

  &:focus-visible {
    outline: none;
  }

  &:disabled {
    cursor: not-allowed;
    opacity: 0.7;
  }

  &--ghost,
  &--transparent {
    color: $primary-color;
    background-color: transparent;

    &:disabled:hover {
      background-color: transparent;
    }
  }

  &--loading {
    min-width: 180px;
  }

  &--icon {
    padding: 0 10px;
    height: 32px;
  }

  &__left-icon,
  &__right-icon {
    display: inline-flex;

    svg {
      height: 16px;
      width: 16px;
    }
  }

  &__left-icon {
    margin-right: 12px;
  }

  &__right-icon {
    margin-left: 12px;
  }
}
