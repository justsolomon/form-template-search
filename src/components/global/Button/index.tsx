import Spinner from '../Spinner';
import styles from './button.module.scss';

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** Classname to apply to button to override default styles */
  className?: string;

  /** Boolean showing the loading state of the button */
  isLoading?: boolean;

  /** Boolean showing the disabled state of the button */
  isDisabled?: boolean;

  /** Determines if the button is an icon button */
  iconButton?: boolean;

  /** Variant of the button. Defaults to `solid` */
  variant?: 'solid' | 'ghost' | 'transparent';

  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  children: React.ReactNode;
}

function Button({
  isDisabled,
  isLoading,
  children,
  className = '',
  iconButton,
  variant = 'solid',
  leftIcon,
  rightIcon,
  ...rest
}: ButtonProps) {
  return (
    <button
      className={`${styles['button']} ${className} ${
        styles['button--' + variant]
      } ${iconButton ? styles['button--icon'] : ''} ${
        isLoading ? styles['button--loading'] : ''
      }`}
      disabled={isDisabled || isLoading}
      data-testid="button"
      {...rest}
    >
      {leftIcon ? (
        <span className={styles['button__left-icon']}>{leftIcon}</span>
      ) : null}

      <>
        {isLoading ? (
          <Spinner size="24px" color="#ffffff" data-testid="button__spinner" />
        ) : (
          children
        )}
      </>

      {rightIcon ? (
        <span className={styles['button__right-icon']}>{rightIcon}</span>
      ) : null}
    </button>
  );
}

export default Button;
