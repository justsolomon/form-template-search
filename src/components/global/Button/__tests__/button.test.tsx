import { render } from '@testing-library/react';
import Button from '..';

describe('<Button />', () => {
  it('should render the button content correctly', () => {
    const { asFragment } = render(<Button>Hey</Button>);

    expect(asFragment).toMatchSnapshot();
  });
});
