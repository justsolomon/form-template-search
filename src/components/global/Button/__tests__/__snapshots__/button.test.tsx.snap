// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the Button component correctly 1`] = `[Function]`;
